### DISCUSSION TITLE
Cleanup data before switching screen

### DISCUSSION THREAD

**Discussion by MuongKimhong** [Textualize/textual#3747] [Q&A] [answered]
Is there anyway to clear all data before I switch to other screen

```python

class Home(Screen):
    data = []

    def compose(self) -> None:
        self.data = fetch_some_data()
        
        yield ScrollableContainer(*self.data, id="container")
        yield Button("Go to Info", id="info-btn")

    def on_button_pressed(self, event: Button.Pressed) -> None:
        if event.button.id == "info-btn":
            self.app.switch_screen("info")

class InfoScreen(Screen):
    def compose(self) -> None:
        yield Static("Info Screen")
        yield Button("Back", id="back-btn)

    def on_button_pressed(self, event: Button.Pressed) -> None:
        if event.button.id == "back-btn":
            self.app.switch_screen("home")

```

For some reason, when I return to HomeScreen I can see that `fetch_some_data()` runs again but the old data inside `data` variable are still there so it just added up new data. How can I clear the old data before I switch the screen?

**Comment by davep** [comment, marked as answer]
It's a little tricky to help here as it's not clear what all if your code is doing, but I would question the suggestion that `fetch_some_data` "runs again". If you're using `switch_screen` to switch between screens, you're going to be switching between existing instances, and so `compose` will only be called once for each screen. You can confirm this with this little example:

```python
from textual.app import App, ComposeResult
from textual.screen import Screen
from textual.widgets import Label

class Info(Screen):

    BINDINGS = [
        ("space", "switch_screen('home')"),
    ]

    def compose(self) -> ComposeResult:
        self.app.bell()
        yield Label("This is the info screen.")


class Home(Screen):

    BINDINGS = [
        ("space", "switch_screen('info')"),
    ]

    def compose(self) -> ComposeResult:
        self.app.bell()
        yield Label("This is the home screen.")

class SwitchScreenApp(App[None]):

    SCREENS = {
        "home": Home(),
        "info": Info(),
    }

    def on_mount(self) -> None:
        self.push_screen("home")

if __name__ == "__main__":
    SwitchScreenApp().run()
```

If `compose` were called on every switch, then you'd hear the terminal bell every time you press space. You will only hear it twice, once each time each screen is first composed.

If you want to do something when a screen is switched, you probably want to be looking at the [`ScreenSuspend`](https://textual.textualize.io/events/screen_suspend/) and [`ScreenResume`](https://textual.textualize.io/events/screen_resume/) events.

**Reply by MuongKimhong** [reply]
ScreenResume and ScreenSuspend solves this. Thanks you!
